### PR TITLE
[FEATURE] - Use a RESTful API architecture instead of heavily relying on Selenium

### PR DESCRIPTION
**Pull request by eyadgaran**
This is probably a controversial change so will leave it here in case admins want to incorporate upstream.

My motivation was I had an extremely difficult time getting the selenium implementation running because of [presumably?] browser and mint ui differences. Selenium is very fickle and depends on particular css and ui code to map to functions. In reality the only thing that is actually needed is an active cookie to inject in request headers. Which brings me to:

I refactored api interaction into a standalone parallel rest client that can be credentialed with a cookie (pulled from the selenium browser or manually pasted from a different browser session). Also noticed there was no inherent pagination scheme so added that as an automatic function for requests

My use case did not extend to the credit endpoints so I left those as is but laid the foundation for what i think would be a relatively low lift refactor to abstract those into generic rest calls as well

### CODE REVIEW

**Comment by mrooney** [comment]
Very cool idea, thanks for sharing! I will check it out :)

**Comment by burkematthew** [comment]
@eyadgaran there is a lot going on here and not in a bad way.  However, I think this should be broken up into multiple feature requests, at a very minimum one that handles implementing pagination and another that implements the RESTful approach.  With as fundamental of a change as this would be, keeping the change as laser-focused as possible is key.  

In general, though, this to me is a major release type of change.  What would be nice is if we could leverage some of the bones implemented in this PR throughout v2 and if we do want to make the switch, we work up to the final implementation\switch in a v3. 

That being said, I'll do some review to get a feel for the overall approach and leave general comments!

**Comment by burkematthew** [comment]
A general question here:

Is the idea that we use Selenium to get us logged in and get our necessary cookies and then we just use the REST client to make all the necessary HTTP requests using those cookies?  Seems pretty awesome and definitely an improvement.

**Comment by eyadgaran** [comment]
thanks for the review @burkematthew (and apologies for the absence)! in a nutshell, yes, the crux of the change is to convert to a standardized REST api with optional browser emulation to extract the api key (the "cookie"). this separation was necessary in my personal usage because selenium issues prevented me from using the current implementation - even though the requests are fully agnostic after authentication. 

I am happy to work on pulling this apart in my spare time to get it incorporated for others, but it may take me a couple of weeks to circle back to (already have my fork running for my project). I'll leave this pr up with the full implementation in case others want to check it out in the meantime and open separate prs for the individual changes.

fwiw the class structure is just a personal preference that i have found helpful for api clients. it has made it easy to compartmentalize user exposed methods with "business logic" from the internal backend routes that are subject to change as the api evolves (refactoring user methods to mirror the api directly can be a pain on breaking changes to syntax/urls, even if content is equivalent)

**Comment by burkematthew** [comment]
@eyadgaran @emann3 I'll work on getting some of the merge conflicts resolved and also see what appears to be the best approach to implementing this style of enhancement. Really feels like more of a major release type of feature, but obviously seems like it brings a ton of value.

**Comment by eyadgaran** [comment]
thanks! ill reference the mini prs here as well as i go through them.

@emann3 i didnt put much effort into refactoring the cli so ill try to take a few minutes later to get that issue resolved for you. cheers

**Comment by snowskeleton** [comment]
I just got this branch working on in my implementation. It's fantastic! The key function went from 20 lines down to 3 lines, and the runtime decreased from 30 seconds to 2 seconds.

**Review by snowskeleton — COMMENTED** [review]
can you explain these changes a little more? PR #549 updated them to keep current with changes Mint made to the signin process.

**Comment by eyadgaran** [comment]
> can you explain these changes a little more? PR #549 updated them to keep current with changes Mint made to the signin process.
> 
> edit: referencing commit [19fb5d1](https://github.com/mintapi/mintapi/pull/482/commits/19fb5d13daa6a089f1face6d4639b49c5f3d4375)

I thought those were merge conflict holdovers. my changes should not touch any of the selenium code so I reverted anything that the pr showed as introducing a change to the main branch

**Comment by eyadgaran** [comment]
> > can you explain these changes a little more? PR #549 updated them to keep current with changes Mint made to the signin process.
> > edit: referencing commit [19fb5d1](https://github.com/mintapi/mintapi/pull/482/commits/19fb5d13daa6a089f1face6d4639b49c5f3d4375)
> 
> I thought those were merge conflict holdovers. my changes should not touch any of the selenium code so I reverted anything that the pr showed as introducing a change to the main branch

looks like git sorted it out as part of this merge https://github.com/mintapi/mintapi/commit/c769926dac13f33a0ffb2bec18a44d999edb73e0

should be in line with the main branch

**Comment by snowskeleton** [comment]
> looks like git sorted it out as part of this merge https://github.com/mintapi/mintapi/commit/c769926dac13f33a0ffb2bec18a44d999edb73e0

indeed, all the right changes look to be included. at least as far as `signIn.py` goes. I'm not familiar with the other files touched by this change, but it doesn't look like anything major.

**Comment by mrooney** [comment]
Closing in favor of #578 , let me know if there's some reason to keep this open as well, however!

**Comment by eyadgaran** [comment]
> Closing in favor of #578 , let me know if there's some reason to keep this open as well, however!

this should have parity now so id prefer to keep this open while i add more updates/tests. @snowskeleton were there additional changes in #578 that i missed?

**Comment by snowskeleton** [comment]
I was able to find at least a couple lines in yours that are more recent than mine as of your most recent commit to eyadgaran/mintapi/rest-api, so I'd say keep this PR open and close #578

**Comment by eyadgaran** [comment]
👍 

fyi changes should be ready for review - it implements the same interface as before unless flagged with `use_rest_client` but uses the new endpoint/pagination abstraction. there are more tests I'd like to add, but nothing that would prevent review

**Review by snowskeleton — DISMISSED** [review]
This all looks great.

**Comment by emann3** [comment]
This works for me also - from the command line (if I set use_rest_client).

Would it be possible to make this a command line switch?

**Comment by snowskeleton** [comment]
Should be fairly simple. In `cli.py` at line 498 add
```
use_rest_client=options.use_rest_client,
```
Then for the cli part something like
```
        (
            ("--use-rest-client",),
            {
                "action": "store_true",
                "help": "Prefer the REST client for API calls. This will be the default in a future release",
            },
        ),
```
in the `ARGUMENTS` dict.

**Review by burkematthew — APPROVED** [review]

File: mintapi/api.py [hunk 1]
```diff
@@ -1,700 +1,93 @@
...
-        tag_ids: List[str] = None,
-        descriptions: List[str] = None,
-        account_ids: List[str] = None,
-        match_all_filters: bool = True,
-        include_investment: bool = False,
-        remove_pending: bool = True,
-        limit: int = 5000,
-        offset: int = 0,
-    ) -> List[Dict]:
+        Automatically handle routing to prefer the rest client but fallback to the browser for uinimplemented
```
[the remark this answers is not in the document]
Comment (mrooney) [reply mintapi/api.py:54]: Good catch, I can fix that post-merge :)

**Review by mrooney — APPROVED** [review]

**Comment by mrooney** [comment]
Woo party time! Thanks for the epic work from everyone involved!!!

**Comment by eyadgaran** [comment]
Woo - thanks team!

**Comment by eyadgaran** [comment]
Still some outstanding tests to implement but I'll open prs as I get to them

**Comment by mrooney** [comment]
@eyadgaran awesome, thanks! There's also https://github.com/snowskeleton/mintapi/pull/8 which, if you like the tests, I'd love to see a PR for on the main mintapi project/branch :)

**Pull request merged**